server/src/main/java/org/opensearch/index/shard/StoreRecovery.java

```diff
@@ -791,4 +795,26 @@ private void bootstrap(final IndexShard indexShard, final Store store) throws IO
         );
         store.associateIndexWithNewTranslog(translogUUID);
     }
+
+    /*
+    Blocks the calling thread,  waiting for the remote store to get synced till internal Remote Upload Timeout
+     */
+    private void waitForRemoteStoreSync(IndexShard indexShard) {
+        if (indexShard.shardRouting.primary() == false) {
+            return;
+        }
+        long startNanos = System.nanoTime();
+
+        while (System.nanoTime() - startNanos < indexShard.getRecoverySettings().internalRemoteUploadTimeout().nanos()) {
+            if (indexShard.isRemoteSegmentStoreInSync()) {
+                break;
+            } else {
+                try {
+                    Thread.sleep(TimeValue.timeValueMinutes(1).seconds());
```

Comment:
since we are blocking the thread for 1 hr in worst case,  is there a way to cancel/stop recovery for the user if they want after waiting for, lets say 15 mins?

Reply:
This doesn't affect the cancel/stop recovery flow as the thread can still be cancelled .